File: Python/graphs.py
from queue import Queue
from _collections import deque

class Graph:
    def __init__(self,V):
        self.V = V
        self.adj = {}

        for i in range(0,V):
            self.adj[i] = []

    def addEdge(self, u, v):
        self.adj[u].append(v)
        self.adj[v].append(u)

    def add_directed_edge(self,u,v):
        self.adj[u].append(v)

    def dfs(self, v):
        marked = [False for i in range(0, self.V)]
        for v in range(0, self.V):
            if not marked[v]:
                self._dfs(v, marked)

    def _dfs(self, v, marked):
        marked[v] = True
        print(v)
        for w in self.adj[v]:
            if not marked[w]:
                self._dfs(w, marked)

    def num_components(self):
        id = {}
        count = 0
        marked = [False for i in range(0, self.V)]

        for v in range(0, self.V):
            if not marked[v]:
                self.dfs_CC(v, id, count, marked)
                count += 1

        return count

    def dfs_CC(self, v, id, count, marked):
        marked[v] = True
        id[v] = count
        print(f'v:{v}, id:{id[v]}')
        for w in self.adj[v]:
            if not marked[w]:
                self.dfs_CC(w, id, count, marked)

    def is_connected(self):
        marked = [False for i in range(0,self.V)]
        self._dfs(0,marked)
        return all(marked)

    def BFS(self,s):
        marked = [False for i in range(0, self.V)]
        q = Queue()
        q.put(s)
        marked[s] = True
        while not q.empty():
            v = q.get()
            print(v)
            for w in self.adj[v]:
                if not marked[w]:
                    q.put(w)
                    marked[w] = True

    def topo_sort(self):
        """
        returns the topological order of the graph
        also called the reversePostOrder
        using deque over list as it provides O(1) operations
        as compared to O(n) for the list

        time: O(|V| + |E|)
        """
        marked = [False for i in range(0, self.V)]
        reversePost = deque()

        for v in range(0,self.V):
            if not marked[v]:
                self._dfs_topo(marked,reversePost,v)

        return reversePost

    def _dfs_topo(self,marked,reversePost,v):
        marked[v] = True
        for w in self.adj[v]:
            if not marked[w]:
                self._dfs_topo(marked,reversePost,w)

        reversePost.append(v)


    def print_topo_order(self):
        reversePost = self.topo_sort()
        print('Topological Order:')
        while reversePost:
            print(reversePost.pop(), end=" ")



if __name__ == "__main__":
    g = Graph(7)
    # g.addEdge(0, 1)
    # g.addEdge(0, 2)
    # g.addEdge(0, 6)
    # g.addEdge(0, 5)
    # g.addEdge(5, 3)
    # g.addEdge(5, 4)
    # g.addEdge(6, 4)
    # g.BFS(0)
    g.add_directed_edge(0, 1)
    g.add_directed_edge(0, 5)
    g.add_directed_edge(0, 2)
    g.add_directed_edge(3, 2)
    g.add_directed_edge(3, 5)
    g.add_directed_edge(5, 2)
    g.add_directed_edge(1, 4)
    g.add_directed_edge(3, 4)
    g.add_directed_edge(3, 6)
    g.add_directed_edge(6, 4)
    g.add_directed_edge(6, 0)
    g.print_topo_order()

File: C++/graphs.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

struct vertex {
    vector<int>adj;
};
class Graph {
protected:
    int V;
    vector<vector<int>> adj;
    queue<int>pre;
    queue<int>post;
    stack<int>reversePost;

public:
    Graph(int V) {
        this->V = V;
        adj.assign(V,vector<int>());

    }

    void dfs(int v) {
        vector<bool>marked(V,false);

        for(int v=0; v<V;v++) {
            if(!marked[v]) {
                _dfs(v,marked);
            }
        }
    }
    void _dfs(int v, vector<bool> &marked) {
        pre.push(v);
        marked[v] = true;
        cout << v << endl;
        for(int w: adj[v]) {
            if (!marked[w])
                _dfs(w,marked);
        }
        post.push(v);
        reversePost.push(v);
    }
    void add_edge(int u, int v) {
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    void add_directed_edge(int v,int w) {
        adj[v].push_back(w);
    }
    stack<int> topo_sort() {
        vector<bool>marked(V, false);
        stack<int>reversePost;

        for(int i=0;i<V;i++) {
            if(!marked[i])
                _dfs_topo(marked,reversePost, i);
        }
        return reversePost;
    }
    void _dfs_topo(vector<bool>&marked,stack<int>&reversePost,int v) {
        marked[v] = true;

        for(int w:adj[v]) {
            if(!marked[w])
            {
                _dfs_topo(marked,reversePost,w);
            }
        }
        reversePost.push(v);
    }
    void print_topo_sort() {
        stack<int>s = this->topo_sort();
        cout << "Topological Order: " << endl;
        while(!s.empty()) {
            cout << s.top() << " ";
            s.pop();
        }
    }
    bool hasCycle()
    {
        vector<bool>marked(V,false);
        vector<bool>on_stack(V,false);
        bool found_cycle = false;
        for(int i=0; i<V; i++)
        {
            if(!marked[i])
                dfs_cycle(i, marked, on_stack, found_cycle);
        }
        return found_cycle;
    }
    void dfs_cycle(int v, vector<bool>&marked, vector<bool>&on_stack, bool& found)
    {
        marked[v] = true;
        on_stack[v] = true;

        for(int w: adj[v])
        {
            if(on_stack[w])
                found = true;
            else
                dfs_cycle(w, marked, on_stack, found);
        }
        on_stack[v] = false;
    }
    int num_components() {
        vector<bool>marked(V,false);
        vector<int>id(V,-1);
        int count=0;

        for(int v=0; v< V;v++) {
            if(!marked[v]) {
                _dfs_CC(marked,id,count,v);
                count++;
            }
        }
        return count;
    }
    void _dfs_CC(vector<bool>&marked,vector<int>&id, int count, int v) {
        marked[v] = true;
        id[v] = count;
        cout << "v:" << v << " id:" << count << endl;
        for(int w:adj[v]) {
            if(!marked[w]) {
                _dfs_CC(marked,id,count,w);
            }
        }
    }
    void preOrder()
    {
        while(!pre.empty()){
            cout << pre.front() << " ";
            pre.pop();
        }
    }
    void postOrder()
    {
        while(!post.empty()){
            cout << post.front() << " ";
            post.pop();
        }
    }
    void reverse_postOrder()
    {
        while(!reversePost.empty()){
            cout << reversePost.top() << " ";
            reversePost.pop();
        }
    }
};


int main() {
    Graph g(7);
//    g.add_edge(0,1);
//    g.add_edge(0,2);
//    g.add_edge(0,6);
//    g.add_edge(0,5);
//    g.add_edge(5,4);
//    g.add_edge(5,3);
//    g.add_edge(6,4);
    g.add_directed_edge(0,1);
    g.add_directed_edge(0,5);
    g.add_directed_edge(0,2);
    g.add_directed_edge(3,2);
    g.add_directed_edge(3,5);
    g.add_directed_edge(5,2);
    g.add_directed_edge(1,4);
    g.add_directed_edge(3,4);
    g.add_directed_edge(3,6);
    g.add_directed_edge(6,4);
    g.add_directed_edge(6,0);
//    g.add_directed_edge(6,3);       // add a cycle 6 - 3
//    cout << g.hasCycle() << endl;
//    g.print_topo_sort();
    g.dfs(0);
    cout << endl;
    g.reverse_postOrder();
    cout << endl;
    g.print_topo_sort();




}